src/shaders.rs

```diff
@@ -305,38 +299,51 @@ pub fn full_shaders(device: &Device, engine: &mut WgpuEngine) -> Result<FullShad
             BindType::Buffer,
         ],
     )?;
-    let fine = match ANTIALIASING {
-        crate::AaConfig::Area => engine.add_shader(
-            device,
-            "fine",
-            preprocess::preprocess(shader!("fine"), &full_config, &imports).into(),
-            &[
-                BindType::Uniform,
-                BindType::BufReadOnly,
-                BindType::BufReadOnly,
-                BindType::BufReadOnly,
-                BindType::Image(ImageFormat::Rgba8),
-                BindType::ImageRead(ImageFormat::Rgba8),
-                BindType::ImageRead(ImageFormat::Rgba8),
-            ],
-        )?,
-        _ => {
-            engine.add_shader(
+
+    let fine_resources = [
+        BindType::Uniform,
+        BindType::BufReadOnly,
+        BindType::BufReadOnly,
+        BindType::BufReadOnly,
+        BindType::Image(ImageFormat::Rgba8),
+        BindType::ImageRead(ImageFormat::Rgba8),
+        BindType::ImageRead(ImageFormat::Rgba8),
+        // Mask LUT buffer, used only when MSAA is enabled.
+        BindType::BufReadOnly,
+    ];
+    let fine_variants = {
+        const AA_MODES: [AaConfig; 3] = [AaConfig::Area, AaConfig::Msaa8, AaConfig::Msaa16];
+        const MSAA_INFO: [Option<(&str, &str)>; 3] = [
+            None,
+            Some(("fine_msaa8", "msaa8")),
+            Some(("fine_msaa16", "msaa16")),
+        ];
+        let mut pipelines = [None, None, None];
+        for (i, aa_mode) in AA_MODES.iter().enumerate() {
+            if options
+                .preferred_antialiasing_method
+                .as_ref()
```

Comment:
Could the `as_ref` be eliminated by comparing `m != *aa_mode`? It's `Copy` for this reason.